### DISCUSSION TITLE
How do I pass and string argument with spaces?

### DISCUSSION THREAD

**Discussion by marcelonyc** [OpenBB-finance/OpenBB#3798] [Q&A]
I need to pass a Date Time argument to a command. The Date time has spaces in the string. How do I handle string arguments that have spaces? (I tried single and double quotes)

 I did not find a datetime validation so I create one in helper_funcs.py.

```
def valid_datetime(s: str) -> datetime:
    """Argparse type to check date is in valid format."""
    try:
        return datetime.strptime(s, "%d/%m/%Y %H:%M %p")
    except ValueError as value_error:
        logging.exception(str(value_error))
        raise argparse.ArgumentTypeError(f"Not a valid datetime: {s}") from value_error
```

When I test my command, the validate_datetime function the parser does not take the full string. 
* If I use **/** to separate **d/m/y h:m** valid_datetime only gets d. 
* If I use **-** to separate d-m-y h:m, valid_datetime gets d-m-y.

![Screenshot from 2022-12-18 08-12-56](https://user-images.githubusercontent.com/12512887/208300540-972b93ee-b282-40e5-a7bb-9db305671c2a.png)

**Comment by northern-64bit** [comment]
Hello @marcelonyc!
The argument parser from argparse only takes arguments without spaces by default, so it only reads the first part of the argument, single and double quotes don't help.
Here is a relevant stackoverflow answer that explains possible solutions well: https://stackoverflow.com/a/15753721

I hope that this helps!

**Comment by jmaslek** [comment]
To followup on @northern-64bit answer, note that `/` breaks the way the terminal works because we allow for paths, so we split on `/`. 
 We can add some regex to allow them, but if they can be avoided,that would be preferred.

Regarding spaces you can do the following:  `nargs=+` which will return a list of what you pass in.  So if you did `--date-time 2022-01-01 22:00 AM`, you would get that ns_parser.datetime = ["2022-01-01", "22:00","AM"], which you could join into a single string then use strptime.

**Reply by marcelonyc** [reply]
@jmaslek I tried nargs=+. But it took any values and parameters entered after as the value of my date time parameter. I think that option will only work if a parameter is that last one. 

For my use case I implemented a valid_datetime in helper_funcs.py. It accepts a DateTime string in any format to be validated and returns in a different format. It will not work with /,but I do not think that is important. My change in helper_funcs is part of PR 3744.

**Comment by markoblogo** [comment]
If you want to keep using argparse while supporting date‑time strings with spaces and avoid breaking other arguments, you essentially have three levers to control:
Use nargs='+' and post‑process
This lets argparse capture the full date‑time sequence, but you must explicitly stop parsing at the right place and join the pieces yourself:
python
parser.add_argument(
    "--date-time",
    dest="date_time",
    nargs="+",
    help="Date-time like '18/12/2022 08:12 AM'",
)
args, unknown = parser.parse_known_args()
if args.date_time:
    dt_str = " ".join(args.date_time)
    dt = datetime.strptime(dt_str, "%d/%m/%Y %H:%M %p")
parse_known_args prevents nargs='+' from greedily swallowing subsequent options, because anything it cannot match as a known option falls into unknown, which you can forward or re‑parse as needed.
Avoid / when the CLI splits on it
As mentioned above, OpenBB’s CLI already splits on / to support paths, so a format like 2022-12-18 08:12 AM is safer than 18/12/2022 08:12 AM. If you really need /, you’d have to alter the higher‑level tokenizer, not just argparse.
Centralize validation/normalization
Your custom valid_datetime helper is a good idea; you can keep it as the single source of truth and accept multiple input patterns:
python
def valid_datetime(s: str) -> datetime:
    for fmt in ["%d/%m/%Y %H:%M %p", "%Y-%m-%d %H:%M %p"]:
        try:
            return datetime.strptime(s, fmt)
        except ValueError:
            continue
    raise argparse.ArgumentTypeError(f"Not a valid datetime: {s}")
Combining this with the nargs='+' + parse_known_args approach gives you robust parsing without changing existing OpenBB behaviour around /.
This way you respect the current CLI design (splitting on /), still support spaces in the date‑time argument, and keep your validation logic reusable and easy to extend.
